include/xtensor/xfunction.hpp

```diff
@@ -768,16 +801,16 @@ namespace xt
     template <class S>
     inline bool xfunction_base<F, R, CT...>::broadcast_shape(S& shape, bool reuse_cache) const
     {
-        if (reuse_cache && m_shape_computed)
+        if(reuse_cache)
         {
-            std::copy(m_shape.cbegin(), m_shape.cend(), shape.begin());
-            return m_shape_trivial;
+            //First call will compute the shape if this has not been done so yet.
+            std::copy(this->shape().cbegin(), this->m_shape.cend(), this->m_shape.begin());
+            return this->m_shape_trivial;
         }
         else
         {
-            // e.broadcast_shape must be evaluated even if b is false
-            auto func = [&shape](bool b, auto&& e) { return e.broadcast_shape(shape) && b; };
-            return accumulate(func, true, m_e);
+            //We only need one promotion operation, and triviality comes fointly from this and the cashed ones.
+            return xt::broadcast_shape(shape, this->shape()) && this->m_shape_trivial;
```

Comment:
We may not want to compute and store the internal shape when broadcasting the expression.